core/src/main/scala/org/http4s/util/UrlCoding.scala

```diff
@@ -44,13 +48,42 @@ private[util] trait UrlCodingUtils {
     })
   }
 
+  /**
+    * Percent-encodes a string.  Depending on the parameters, this method is
+    * appropriate for URI or URL form encoding.  Any resulting percent-encodings
+    * are normalized to uppercase.
+    * 
+    * Characeters in `toSkip` are rendered as is.  If a skipped '%' is followed
+    * by exactly two hex digits, they are uppercased for normalization purposes.
+    * 
+    * Space is rendered as `" "` if skipped.  If space is not skipped, it is
+    * encoded as `"+"` when `spaceIsPlus` is true, or percent encoded otherwise.
+    * 
+    * All characters not in toSkip are rendered in a percent encoding, according
+    * to `charset`.
+    */
   def urlEncode(toEncode: String, charset: Charset = Utf8, spaceIsPlus: Boolean = false, toSkip: BitSet = toSkip) = {
-    val in = charset.encode(ensureUppercasedEncodings(toEncode))
```

Comment:
I think the `ensureUppercasedEncodings` is what doesn't belong here: `toEncode` should be the raw input so we shouldn't be perturbing it.